respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,327 @@
+       Journal of Econometrics, 47(2-3), 197-205.
+.. [3] Duffie, D., & Singleton, K. (1993). Simulated Moments Estimation of Markov Models
+       of Asset Prices. Econometrica, 61(4), 929-952.
+"""
+import copy
+import functools
+
+import numpy as np
+import pandas as pd
+
+from respy.simulate import get_simulate_func
+
+
+def _harmonize_input(arg):
+    """ Harmonizes different types of inputs by turning all inputs into lists.
+    - pandas.DataFrames/Series and callable functions will turn into list
+        containing a single item (i.e. the input).
+    - Dictionaries will be sorted according to keys and then turn into a list
+        containing the dictionary entries.
+    """
+    # Convert single DataFrames, Series or function into list containing one
+    # item.
+    if isinstance(arg, (pd.DataFrame, pd.Series)) | callable(arg):
+        arg = [arg]
+
+    # Sort dictionary accoding to keys and turn into list.
+    if isinstance(arg, dict):
+        temp = []
+        for key in sorted(arg.keys()):
+            temp.append(arg[key])
+        arg = temp
+
+    # Leave lists as is.
+    if isinstance(arg, list):
+        pass
+    else:
+        raise TypeError(
+            "Function only accepts lists, dictionaries,"
+            " functions, Series and DataFrames as inputs."
+        )
+
+    return arg
+
+
+def _calculate_moments(data, calc_moments):
+    """ Uses functions specified in calc_moments to calculate moments for the
+    data.
+    """
+    data = data.copy()
+    moments = []
+    for func in calc_moments:
+        moments.append(func(data))
+
+    return moments
+
+
+def _reindex_simulated_moments(empirical_moments, simulated_moments):
+    """ Reindexes data in simulated_moments to match index of
+    empirical_moments.
+    """
+    simulated_moments_reindexed = []
+
+    for emp_mom, sim_mom in zip(empirical_moments, simulated_moments):
+        col = emp_mom.columns
+        sim_mom = sim_mom.reindex(columns=col)
+
+        simulated_moments_reindexed.append(sim_mom)
+
+    return simulated_moments_reindexed
+
+
+def _replace_missing_values(simulated_moments, replace_nans):
+    """ Replaces missing values in simulated_moments according to replacement
+    functions specified in replace_nans.
+    """
+    moments = []
+
+    for mom, func in zip(simulated_moments, replace_nans):
+        mom = func(mom)
+        moments.append(mom)
+
+    return moments
+
+
+def _flatten_index(data):
+    """ Function flattens index of DataFrames/Series contained in data and
+        concatinates all flattend elements in the list to one long DataFrame.
+    """
+    data_flat = []
+    for df in data:
+        # Unstack df to add columns to index and extract indexes from the
+        # new df.
+        df.index = df.index.map(str)
+        df = df.unstack()
+        index = pd.MultiIndex.to_flat_index(df.index)
+
+        # Flatten index
+        index_flat = []
+        for idx in index:
```

Comment:
I think you can use ``index.str.join("_")`` ([link](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.join.html)) here or even instead of ``index = pd.MultiIndex.to_flat_index(df.index)``. Unstacking is a clever idea 👍.

Reply:
Doesn't work on MultiIndex but this still reduced a chunk of the code to a single line. Thanks for the tip!